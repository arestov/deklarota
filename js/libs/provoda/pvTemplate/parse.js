
import getCachedPVData from './getCachedPVData'

function getCommentPVData(cur_node, struc_store, getSample) {
  return getCachedPVData(cur_node, struc_store, true, getSample)
}

function getPVData(cur_node, struc_store, getSample) {
  return getCachedPVData(cur_node, struc_store, false, getSample)
}

export default function parse(start_node, struc_store, getSample, _opts) {
  //полный парсинг, без байндинга

  var result = []

  var match_stack = [ start_node ]
  var i = 0
  while (match_stack.length) {
    var cur_node = match_stack.shift()
    var node_type = cur_node.nodeType
    var directives_data = null
    if (node_type == 1) {
      directives_data = getPVData(cur_node, struc_store, getSample)
      result.push(cur_node, directives_data)
    } else if (node_type == 8) {
      directives_data = getCommentPVData(cur_node, struc_store, getSample)
      result.push(cur_node, directives_data)
    }

    // if (directives_data.replacer) {
    // 	match_stack.unshift(directives_data.node);
    // }

    if (node_type == 1) {
      for (i = 0; i < cur_node.childNodes.length; i++) {
        match_stack.push(cur_node.childNodes[i])
      }
    }

  }
  return result
}
