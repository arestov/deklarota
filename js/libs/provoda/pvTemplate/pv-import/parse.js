

export default function parsePVImport(node, sample_name) {

  var possible = node.nodeName === 'SCRIPT'
    ? node
    : node.querySelector('script[type="pv-import-map"]')

  var script
  if (possible === node) {
    script = node
  }

  if (possible.parentNode === node) {
    script = node
    node.removeChild(script)
  }
  var map_string = script && script.textContent
  var map = map_string ? JSON.parse(map_string) : [{}]

  return {
    sample_name: sample_name,
    map: map,
    pv_nest: script.getAttribute('pv-nest') || null
  }
};
