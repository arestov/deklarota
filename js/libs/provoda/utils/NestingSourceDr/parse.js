
import spv from '../../../spv'
const memorize = spv.memorize
const splitByDot = spv.splitByDot

const NestingSourceDr = function(string) {
  const parts = string.split('>')
  this.start_point = parts.length > 1 && parts[0]
  this.selector = splitByDot(parts[parts.length - 1])
}

export default memorize(function parseNSD(string) {
  return new NestingSourceDr(string)
})
