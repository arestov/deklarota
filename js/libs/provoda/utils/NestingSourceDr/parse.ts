
import memorize from '../../../spv/memorize'
import splitByDot from '../../../spv/splitByDot'
import type { RelPath } from '../multiPath/addr.types'

type NestingSource = {
  start_point: string | false
  selector: RelPath
}

class NestingSourceDr {
  start_point: string | false
  selector: RelPath
  constructor(string: string) {
    const parts = string.split('>')
    this.start_point = parts.length > 1 && (parts[0] || false)
    const last_part = parts[parts.length - 1]
    if (last_part == null) {
      throw new Error('provide last part of path')
    }
    this.selector = splitByDot(last_part)
  }
}

export default memorize(function parseNSD(string) {
  return new NestingSourceDr(string) as NestingSource
})
