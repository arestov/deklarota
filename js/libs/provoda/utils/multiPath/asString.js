

export default function multiPathAsString(multi_path) {
  if (multi_path.as_string) {
    return multi_path.as_string
  }

  multi_path.as_string = ''
    + firstPart(isStateOk(multi_path.state) && multi_path.zip_name, multi_path.state)
    + nestingString(!isStateOk(multi_path.state) && multi_path.zip_name, multi_path.nesting)
    + resourceString(multi_path.resource)
    + baseString(multi_path.from_base)

  return multi_path.as_string
};

function isStateOk(state) {
  return state && state.path
}

function wrapBySpace(item) {
  if (!item) {
    return ''
  }

  return ' ' + item + ' '
}

function firstPart(zip_name, state) {
  return '<' + wrapBySpace(zipPart(zip_name) + stateString(state))
}

function zipPart(zip_name) {
  if (!zip_name) {
    return ''
  }

  return '@' + zip_name + ':'
}


function stateString(state) {
  if (!isStateOk(state)) {
    return ''
  }

  return state.path
}

function isNestingOk(nesting) {
  return nesting && nesting.path
}

function nestingString(zip_name, nesting) {
  if (!isNestingOk(nesting)) {
    return '<'
  }

  var path = nesting.path.join('.')

  return '<' + wrapBySpace(zipPart(zip_name) + path)
}

function resourceString(resource) {
  if (!resource || !resource.path) {
    return '<'
  }

  return '< ' + resource.path + ' '
}

function baseStringMin(from_base) {
  if (!from_base || !from_base.type) {
    return ''
  }

  switch (from_base.type) {
    case 'root': {
      return '#'
    }
    case 'parent': {
      var repeated = ''
      var counter = 1
      while (counter <= from_base.steps) {
        repeated += '^'
        counter++
      }
      return repeated
    }

  }
}

function baseString(from_base) {
  var result = baseStringMin(from_base)
  return result ? '< ' + result : '<'
}

export { baseStringMin }
