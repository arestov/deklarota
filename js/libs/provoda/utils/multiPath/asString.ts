import type { EmptyObject } from '../empty.types'
import type { ParentAscendor, RootAscendor } from './addr-parts/ascendor.types'
import type { AttrPartOfAddr } from './addr-parts/attr.types'
import type { RelPartOfAddr } from './addr-parts/rel.types'
import type { RoutePartOfAddr } from './addr-parts/routes.types'

type Ascendor = RootAscendor | ParentAscendor | EmptyObject
type Route = RoutePartOfAddr | EmptyObject
type Rel = RelPartOfAddr | EmptyObject
type Zip = null | string

type Addr = {
  zip_name: Zip
  as_string: null | string
  state: AttrPartOfAddr
  nesting: Rel
  resource: Route
  from_base: Ascendor
}

type ZipArg = Zip | boolean


export default function multiPathAsString(multi_path: Addr): string {
  if (multi_path.as_string) {
    return multi_path.as_string
  }

  multi_path.as_string = ''
    + firstPart(isStateOk(multi_path.state) && multi_path.zip_name, multi_path.state)
    + nestingString(!isStateOk(multi_path.state) && multi_path.zip_name, multi_path.nesting)
    + resourceString(multi_path.resource)
    + baseString(multi_path.from_base)

  return multi_path.as_string
};

function isStateOk(state: AttrPartOfAddr): boolean {
  return Boolean(state && state.path)
}

function wrapBySpace(item: string): string {
  if (!item) {
    return ''
  }

  return ' ' + item + ' '
}

function firstPart(zip_name: ZipArg, state: AttrPartOfAddr): string {
  return '<' + wrapBySpace(zipPart(zip_name) + stateString(state))
}

function zipPart(zip_name: ZipArg): string {
  if (!zip_name) {
    return ''
  }

  return '@' + zip_name + ':'
}


function stateString(state: AttrPartOfAddr): string {
  if (!isStateOk(state)) {
    return ''
  }

  return state.path
}

function isNestingOk(nesting: Rel): boolean {
  return Boolean(nesting && nesting.path)
}

function nestingString(zip_name: ZipArg, nesting: Rel): string {
  if (!isNestingOk(nesting)) {
    return '<'
  }

  const path = nesting.path.join('.')

  return '<' + wrapBySpace(zipPart(zip_name) + path)
}

function resourceString(resource: Route): string {
  if (!resource || !resource.path) {
    return '<'
  }

  return '< ' + resource.path + ' '
}

function baseStringMin(from_base: Ascendor): string {
  if (!from_base || !from_base.type) {
    return ''
  }

  switch (from_base.type) {
    case 'root': {
      return '#'
    }
    case 'parent': {
      let repeated = ''
      let counter = 1
      while (counter <= from_base.steps) {
        repeated += '^'
        counter++
      }
      return repeated
    }

  }
}

function baseString(from_base: Ascendor): string {
  const result = baseStringMin(from_base)
  return result ? '< ' + result : '<'
}

export { baseStringMin }
