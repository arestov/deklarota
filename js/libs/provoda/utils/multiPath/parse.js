

import spv from '../../../spv'
import getParsedPath from '../../routes/legacy/getParsedPath'
import supportedZip from './supportedZip'
import fromLegacy from './fromLegacy'
import isJustAttrAddr from './isJustAttrAddr'

var splitByDot = spv.splitByDot
var empty = Object.freeze({})
var root = Object.freeze({
  type: 'root',
  steps: null,
})
var parents = spv.memorize(function(num) {
  return Object.freeze({
    type: 'parent',
    steps: num,
  })
})
var parent_count_regexp = /\^+/gi

/*


/(\^|\s+)(\<)(\s+)/
"< @all state_name < nesting < resource < #"

"< state_name < aggr:nesting < resource < #"

"< state_name << /resource/[:ddaf]/sdf < #"

"< state_name <<< #"
"<< nesting_name << #"

"<< nesting_name << ^^"
"< state_name <<< ^^"

"< state_name"
"state_name"

"/resource/[:ddaf]/sdf < #"
"/resource/[:ddaf]/sdf <"

"nesting_name < < ^^"
*/
var checkSplit = /(?:^|\s+)?<(?:\s+)?/
var end = /(<$)|(\^$)|(#$)/
var start = /^</

var parseFromStart = spv.memorize(function(string) {
  var parts = string.split(checkSplit)
  // parts[0] should be empty
  var state = parts[1]
  var nest = parts[2]
  var resource = parts[3]
  var base = parts[4]

  return parseParts(state, nest, resource, base)

})

var parseFromEnd = spv.memorize(function(string) {
  var parts = string.split(checkSplit)

  var length = parts.length
  var base = parts[length - 1]
  var resource = parts[length - 2]
  var nest = parts[length - 3]
  var state = parts[length - 4]

  return parseParts(state, nest, resource, base)
})

function canParseModern(string) {
  var from_start = start.test(string)
  var from_end = end.test(string)
  return (from_start || from_end)
    ? {from_start: from_start, from_end: from_end}
    : null
}

var parseModern = spv.memorize(function parseModern(string) {
  var can_parse = canParseModern(string)
  if (can_parse == null) {
    return null
  }

  if (can_parse.from_start) {
    return parseFromStart(string)
  }
  return parseFromEnd(string)
})

var matchNotStateSymbols = /(^\W)|\@|\:/


export var getStateInfo = spv.memorize(function getStateInfo(string) {
  if (!string) {
    return empty
  }

  return {
    base: splitByDot(string)[0],
    path: string,
  }
})

var SimpleStateAddr = function(string) {
  this.state = getStateInfo(string)
}

SimpleStateAddr.prototype = spv.cloneObj(SimpleStateAddr.prototype, {
  result_type: 'state',
  zip_name: null,
  state: null,
  nesting: empty,
  resource: empty,
  from_base: empty,
  as_string: null,
})

var simpleState = spv.memorize(function simpleState(string) {
  return new SimpleStateAddr(string)
})

var attemptSimpleStateName = function(string) {
  if (!string || matchNotStateSymbols.test(string)) {
    return null
  }

  return simpleState(string)
}

var self = {
  as_string: '<<<<',
  base_itself: true,
}

var parseMultiPath = function(string, allow_legacy) {
  if (string == '<<<<') {
    return self
  }


  var modern = parseModern(string)
  if (modern != null) {
    var state_info = modern.state
    if (state_info && state_info.base && !isJustAttrAddr(modern) && state_info.base.startsWith('__')) {
      console.warn(new Error('private attr used'), string)
    }

    return modern
  }

  return attemptSimpleStateName(string) || (
    allow_legacy ? fromLegacy(string) : null
  )

}
var matchZip = /(?:\@(.+?)\:)?(.+)?/

var parseLegacyCached = spv.memorize(parseMultiPath)
var parseModernCached = spv.memorize(parseMultiPath)


var parseWithCache = function(addr_str, legacy_ok) {
  if (legacy_ok != true) {
    return parseModernCached(addr_str, false)
  }

  return parseLegacyCached(addr_str, true)
}


parseWithCache.simpleState = simpleState

export default parseWithCache

function parseParts(state_raw, nest_raw, resource_raw, base_raw) {
  var state_part_splited = state_raw && state_raw.match(matchZip)
  var zip_state_string = state_part_splited && state_part_splited[1]
  var state_string = state_part_splited && state_part_splited[2]

  var nest_part_splited = nest_raw && nest_raw.match(matchZip)
  var zip_nest_string = nest_part_splited && nest_part_splited[1]
  var nest_string = nest_part_splited && nest_part_splited[2]

  if (zip_state_string && zip_nest_string) {
    throw new Error('both state and nesting cant have zip_name')
  }

  if (zip_nest_string && state_string) {
    throw new Error('use state zip')
  }

  if (zip_state_string && !state_string && nest_string) {
    throw new Error('use nest zip')
  }

  if (zip_state_string && !supportedZip(zip_state_string, 'state')) {
    throw new Error('unsupported zip for state: ' + zip_state_string)
  }

  if (zip_nest_string && !supportedZip(zip_nest_string, 'nesting')) {
    throw new Error('unsupported zip for state: ' + zip_nest_string)
  }


  var zip_name = zip_state_string || zip_nest_string || null
  var state_info = getStateInfo(state_string)
  var nest_info = getNestInfo(nest_string)
  var resource_info = getResourceInfo(resource_raw)
  var base_info = getBaseInfo(base_raw)

  var result_type = getResultType(state_info, nest_info, resource_info, base_info)

  return {
    result_type: result_type,
    zip_name: zip_name,
    state: state_info,
    nesting: nest_info,
    resource: resource_info,
    from_base: base_info,
    as_string: null,
  }
}


export function updateResultType(draft) {
  draft.result_type = getResultType(draft.state, draft.nesting, draft.resource, draft.from_base)

}

export function createAddrByPart(source) {
  var draft = {
    result_type: null,
    zip_name: null,
    state: source.state || empty ,
    nesting: source.nesting || empty,
    resource: source.resource || empty,
    from_base: source.from_base || empty,
    as_string: null,
  }

  updateResultType(draft)
  return draft
}


export function getNestInfo(string) {
  if (!string) {
    return empty
  }

  var parts = string.split(':')
  var path = parts.pop()

  var full_path = splitByDot(path)

  var zip_name = parts[0] || null

  if (zip_name) {
    throw new Error('dont use. use < @[zip_name] [statename] < [nestingname]')
  }

  var target_nest_name = full_path[full_path.length - 1] // last one

  if (!target_nest_name) {
    throw new Error('wrong nest path: ' + string)
  }

  return {
    path: full_path,
    base: full_path.slice(0, full_path.length - 1), // all, except last
    target_nest_name: target_nest_name,
  }
}

export function getResourceInfo(string) {
  if (!string) {
    return empty
  }

  if (string.startsWith('#')) {
    throw new Error('use "ascending part" for root/parent traversing')
  }

  return {
    path: string,
    template: getParsedPath(string),
  }
}

export function getBaseInfo(string) {
  if (!string) {
    return empty
  }

  if (string == '#') {
    return root
  }

  var from_parent_num = string.match(parent_count_regexp)
  if (from_parent_num) {
    return parents(from_parent_num[0].length)
  }

  throw new Error('unsupported base: ' + string)
}

function getResultType(state, nest) {
  if (state && state.path) {
    return 'state'
  }

  if (nest && nest.path) {
    return 'nesting'
  }

  return null
}
