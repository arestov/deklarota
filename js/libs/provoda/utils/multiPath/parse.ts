

import supportedZip from './supportedZip'
import fromLegacy from './fromLegacy'
import isJustAttrAddr from './isJustAttrAddr'
import memorize from '../../../spv/memorize'
import parseAttrPart from './addr-parts/attr'
import { parents, parseAscendorPart } from './addr-parts/ascendor'
import parseRelPart from './addr-parts/rel'
import parseRoutePart from './addr-parts/route'
import { emptyObject } from '../sameObjectIfEmpty'
import type { Addr, AddrResultKind, AddrSelf, AscendorAddr, AttrAddr, RelAddr, RouteAddr } from './addr.types'

export { parseAscendorPart as getBaseInfo }
export { parseRoutePart as getResourceInfo }

type AddrDraft = Partial<Addr>

const empty = emptyObject

/*


/(\^|\s+)(\<)(\s+)/
"< @all state_name < nesting < resource < #"

"< state_name < aggr:nesting < resource < #"

"< state_name << /resource/[:ddaf]/sdf < #"

"< state_name <<< #"
"<< nesting_name << #"

"<< nesting_name << ^^"
"< state_name <<< ^^"

"< state_name"
"state_name"

"/resource/[:ddaf]/sdf < #"
"/resource/[:ddaf]/sdf <"

"nesting_name < < ^^"
*/
const checkSplit = /(?:^|\s+)?<(?:\s+)?/
const end = /(<$)|(\^$)|(#$)/
const start = /^</

const parseFromStart = memorize(function(string: string) {
  const parts = string.split(checkSplit)
  // parts[0] should be empty
  const state = parts[1]
  const nest = parts[2]
  const resource = parts[3]
  const base = parts[4]

  return parseParts(state, nest, resource, base)

})

const parseFromEnd = memorize(function(string) {
  const parts = string.split(checkSplit)

  const length = parts.length
  const base = parts[length - 1]
  const resource = parts[length - 2]
  const nest = parts[length - 3]
  const state = parts[length - 4]

  return parseParts(state, nest, resource, base)
})

const parseModern = memorize(function parseModern(string: string): Addr | null {
  if (start.test(string)) {
    return parseFromStart(string)
  }
  if (end.test(string)) {
    return parseFromEnd(string)
  }

  return null
})

const matchNotStateSymbols = /(^\W)|\@|\:/


export const getStateInfo = memorize(parseAttrPart)

type SimpleAttrAddr = {
  // eslint-disable-next-line no-unused-vars
  new (string: string): Addr
}

// eslint-disable-next-line no-unused-vars
const SimpleStateAddr = function(this: Addr, string: string): void {
  this.state = getStateInfo(string)
} as unknown as SimpleAttrAddr

Object.assign(SimpleStateAddr.prototype, {
  result_type: 'state',
  zip_name: null,
  state: null,
  nesting: empty,
  resource: empty,
  from_base: empty,
  as_string: null,
})

const simpleState = memorize(function simpleState(string: string): Addr {
  const result = new SimpleStateAddr(string)
  return result
})

const attemptSimpleStateName = function(string: string): null | Addr {
  if (!string || matchNotStateSymbols.test(string)) {
    return null
  }

  return simpleState(string)
}

const self: AddrSelf = Object.freeze({
  as_string: '<<<<',
  base_itself: true,
})


export const getNestInfo = memorize(parseRelPart)

const parseMultiPath = function(string: string, allow_legacy?: boolean): Addr | AddrSelf | null {
  if (string == '<<<<') {
    return self
  }


  const modern = parseModern(string)
  if (modern != null) {
    const state_info = modern.state
    if (state_info && state_info.base && !isJustAttrAddr(modern) && state_info.base.startsWith('__')) {
      console.warn(new Error('private attr used'), string)
    }

    return modern
  }

  return attemptSimpleStateName(string) || (
    allow_legacy ? fromLegacy(string) : null
  )

}
const matchZip = /(?:\@(.+?)\:)?(.+)?/

const parseLegacyCached = memorize(parseMultiPath)
const parseModernCached = memorize(parseMultiPath)


const parseWithCache = function(addr_str: string, legacy_ok?: boolean): Addr | AddrSelf | null {
  if (legacy_ok != true) {
    return parseModernCached(addr_str, false)
  }

  return parseLegacyCached(addr_str, true)
}


parseWithCache.simpleState = simpleState

export default parseWithCache

function migrateRelString(rel_string: string | undefined | null, base: AscendorAddr, resource: RouteAddr): string | undefined | null {
  if (!base.type) {
    return rel_string
  }

  if (resource.path) {
    return rel_string
  }

  const path_end = rel_string ? `.${rel_string}` : ''

  switch (base.type) {
    case 'root':
      return `$root${path_end}`
    case 'parent': {
      const full_prefix_path: string[] = []
      for (let i = 0; i < base.steps; i++) {
        full_prefix_path.push('$parent')
      }
      return `${full_prefix_path.join('.')}${path_end}`
    }
  }
}

function parseParts(state_raw?: string, nest_raw?: string, resource_raw?: string, base_raw?: string): Addr {
  const state_part_splited = state_raw && state_raw.match(matchZip)
  const zip_state_string = state_part_splited && state_part_splited[1]
  const state_string = state_part_splited && state_part_splited[2]

  const nest_part_splited = nest_raw && nest_raw.match(matchZip)
  const zip_nest_string = nest_part_splited && nest_part_splited[1]
  const nest_string = nest_part_splited && nest_part_splited[2]

  if (zip_state_string && zip_nest_string) {
    throw new Error('both state and nesting cant have zip_name')
  }

  if (zip_nest_string && state_string) {
    throw new Error('use state zip')
  }

  if (zip_state_string && !state_string && nest_string) {
    throw new Error('use nest zip')
  }

  if (zip_state_string && !supportedZip(zip_state_string, 'state')) {
    throw new Error('unsupported zip for state: ' + zip_state_string)
  }

  if (zip_nest_string && !supportedZip(zip_nest_string, 'nesting')) {
    throw new Error('unsupported zip for state: ' + zip_nest_string)
  }
  const base_info = parseAscendorPart(base_raw || null)
  const resource_info = parseRoutePart(resource_raw || null)

  const migrated_rel_string = migrateRelString(nest_string, base_info, resource_info)

  const zip_name = zip_state_string || zip_nest_string || null || (migrated_rel_string != nest_string ? 'one' : null)
  const state_info = getStateInfo(state_string || null)
  const nest_info = getNestInfo(migrated_rel_string || null)

  const result_type = getResultType(state_info, nest_info)

  return Object.seal({
    result_type: result_type,
    zip_name: zip_name,
    state: state_info,
    nesting: nest_info,
    resource: resource_info,
    from_base: resource_info.path ? base_info : emptyObject,
    as_string: null,
  })
}


export function updateResultType(draft: Addr): void {
  draft.result_type = getResultType(draft.state, draft.nesting)

}

export function createAddrByPart(source: AddrDraft): Addr {
  const draft = Object.seal({
    result_type: null,
    zip_name: source.zip_name || null,
    state: source.state || empty ,
    nesting: source.nesting || empty,
    resource: source.resource || empty,
    from_base: source.from_base || empty,
    as_string: null,
  })

  updateResultType(draft)
  return draft
}

function getResultType(state: AttrAddr, nest: RelAddr): AddrResultKind {
  if (state && state.path) {
    return 'state'
  }

  if (nest && nest.path) {
    return 'nesting'
  }

  return null
}

export const clearCache = (): void => {
  parents.__clear()
  parseFromStart.__clear()
  parseFromEnd.__clear()
  parseModern.__clear()
  getStateInfo.__clear()
  simpleState.__clear()
  getNestInfo.__clear()
  parseLegacyCached.__clear()
  parseModernCached.__clear()
}
