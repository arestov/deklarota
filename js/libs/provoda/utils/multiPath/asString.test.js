import parse from './parse'
import asString from './asString'

describe('asString', () => {
  describe('called on any valid addr', () => {
    describe('with `as_string` set', () => {
      it('should return `as_string` property value', () => {
        expect(asString({ as_string: '< enabled <<< #' })).toEqual('< enabled <<< #')
      })
    })

    describe('with `as_string` unset', () => {
      it('should cache result into `as_string` property', () => {
        const parsed = parse('trivialAttr')
        const serialized = asString(parsed)
        expect(parsed.as_string).toEqual(serialized)
      })
    })
  })

  describe('smoke', () => {
    it.each([
      ['root route attr', [
        '< @one:state_name < nesting < resource < #',
        '< @one:state_name < nesting < resource < #',
      ]],
      ['nesting shortcut', [
        '<< @one:nesting',
        '<< @one:nesting <<',
      ]],
      ['full root query', [
        '< state_name < nesting < resource < #',
        '< state_name < nesting < resource < #',
      ]],
      ['full root query with route', [
        '< state_name << /resource/[:ddaf]/sdf < #',
        '< state_name << /resource/[:ddaf]/sdf < #',
      ]],
      ['root attr query', [
        '< state_name <<< #',
        '< state_name <<< #',
      ]],
      ['root nesting query', [
        '<< nesting_name << #',
        '<< nesting_name << #',
      ]],
      ['grandparent nesting query', [
        '<< nesting_name << ^^',
        '<< nesting_name << ^^',
      ]],
      ['grandparent attr query', [
        '< state_name <<< ^^',
        '< state_name <<< ^^',
      ]],
      ['attr query shortcut', [
        '< state_name',
        '< state_name <<<',
      ]],
      ['trivial attr query', [
        'state_name',
        '< state_name <<<',
      ]],
      ['root route query shortcut', [
        '/resource/[:ddaf]/sdf < #',
        '<<< /resource/[:ddaf]/sdf < #'
      ]],
      ['route shortcut', [
        '/resource/[:ddaf]/sdf <',
        '<<< /resource/[:ddaf]/sdf <',
      ]],
      ['grandparent nesting shortcut', [
        'nesting_name << ^^',
        '<< nesting_name << ^^'
      ]]
    ])('should work for %s', (title, [input, expected]) => {
      expect(asString(parse(input))).toEqual(expected)
    })
  })
})
