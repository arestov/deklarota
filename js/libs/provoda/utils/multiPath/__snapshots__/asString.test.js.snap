// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asString smoke should work for attr query shortcut: < state_name <<< 1`] = `"< state_name <<<"`;

exports[`asString smoke should work for full root query with route: < state_name << resource/[:ddaf]/sdf < # 1`] = `"< state_name << resource/[:ddaf]/sdf < #"`;

exports[`asString smoke should work for full root query: < state_name < nesting < resource < # 1`] = `"< state_name < nesting < resource < #"`;

exports[`asString smoke should work for grandparent attr query: < state_name <<< ^^ 1`] = `"< @one:state_name < $parent.$parent <<"`;

exports[`asString smoke should work for grandparent nesting query: << nesting_name << ^^ 1`] = `"<< @one:$parent.$parent.nesting_name <<"`;

exports[`asString smoke should work for grandparent nesting shortcut: << nesting_name << ^^ 1`] = `"<< @one:$parent.$parent.nesting_name <<"`;

exports[`asString smoke should work for nesting shortcut: << @one:nesting << 1`] = `"<< @one:nesting <<"`;

exports[`asString smoke should work for root attr query: < state_name <<< # 1`] = `"< @one:state_name < $root <<"`;

exports[`asString smoke should work for root nesting query: << nesting_name << # 1`] = `"<< @one:$root.nesting_name <<"`;

exports[`asString smoke should work for root route attr: < @one:state_name < nesting < resource < # 1`] = `"< @one:state_name < nesting < resource < #"`;

exports[`asString smoke should work for root route query shortcut: <<< resource/[:ddaf]/sdf < # 1`] = `"<<< resource/[:ddaf]/sdf < #"`;

exports[`asString smoke should work for route shortcut: <<< resource/[:ddaf]/sdf < 1`] = `"<<< resource/[:ddaf]/sdf <"`;

exports[`asString smoke should work for trivial attr query: < state_name <<< 1`] = `"< state_name <<<"`;
